client/state/utils/reducer-utils.js

```diff
@@ -0,0 +1,254 @@
+ * the hasCustomPersistence property to true on the reducer.
+ *
+ * @example
+ * const date = ( state = new Date( 0 ), action ) => {
+ * 	switch ( action.type ) {
+ * 		case 'GROW':
+ * 			return new Date( state.getTime() + 1 );
+ * 		case SERIALIZE:
+ * 			return state.getTime();
+ * 		case DESERIALIZE:
+ * 			if ( isValidStateWithSchema( state, schema ) ) {
+ * 				return new Date( state );
+ * 			}
+ * 			return new Date( 0 );
+ * 		default:
+ * 			return state;
+ * 	}
+ * };
+ * date.hasCustomPersistence = true;
+ *
+ * const combinedReducer = combineReducers( {
+ *     date,
+ *     height
+ * } );
+ *
+ * combinedReducer( { date: -5, height: -5 } ), { type: DESERIALIZE } ); // { date: new Date( 0 ), height: 150 };
+ * combinedReducer( { date: -5, height: 123 } ), { type: DESERIALIZE } ); // { date: new Date( 0 ), height: 150 };
+ * combinedReducer( { date:  6, height: 123 } ), { type: DESERIALIZE } ); // { date: new Date( 6 ), height: 150 };
+ * combinedReducer( { date: new Date( 6 ), height: 123 } ), { type: SERIALIZE } ); // { date: 6, height: 150 };
+ * combinedReducer( { date: new Date( 6 ), height: 123 } ), { type: GROW } ); // { date: new Date( 7 ), height: 124 };
+ *
+ * @param {object} reducers - object containing the reducers to merge
+ * @returns {function} - Returns the combined reducer function
+ */
+export function combineReducers( reducers ) {
+	// set up persistence of reducers passed from app and then create a combined one
+	return createCombinedReducer( mapValues( reducers, setupReducerPersistence ) );
+}
+
+function applyStoredState( reducers, state, action ) {
+	let hasChanged = false;
+	const nextState = mapValues( reducers, ( reducer, key ) => {
+		// Replace the value for the key we want to init with action.storedState.
+		if ( reducer.storageKey === action.storageKey ) {
+			hasChanged = true;
+			return action.storedState;
+		}
+
+		// Descend into nested state levels, possibly the storageKey will be found there?
+		const prevStateForKey = state[ key ];
+		const nextStateForKey = reducer( prevStateForKey, action );
+		hasChanged = hasChanged || nextStateForKey !== prevStateForKey;
+		return nextStateForKey;
+	} );
+
+	// return identical state if the stored state didn't get applied in this reducer
+	return hasChanged ? nextState : state;
+}
+
+function createCombinedReducer( reducers ) {
+	const combined = combine( reducers );
+
+	const combinedReducer = ( state, action ) => {
+		switch ( action.type ) {
+			case SERIALIZE:
+				return serializeState( reducers, state, action );
+
+			case APPLY_STORED_STATE:
+				return applyStoredState( reducers, state, action );
+
+			default:
+				return combined( state, action );
+		}
+	};
+
+	combinedReducer.hasCustomPersistence = true;
+	combinedReducer.addReducer = addReducer( combinedReducer, reducers );
+	combinedReducer.getStorageKeys = getStorageKeys( reducers );
+
+	return combinedReducer;
+}
+
+function getStorageKeys( reducers ) {
```

Comment:
Maybe this locally used util should live by `withStorageKey`?

Reply:
Not sure I have a great grasp of both utils, but judging from a quick look, `getStorageKeys` is used in `client/state/initial-state` (to determine storage keys to load persisted state), whereas `withStorageKey` seems to be used only by WooCommerce (and Zoninator). So since their domains are fairly different, I think it's fine to keep their definitions in separate places (and maybe look into whether we can phase out `withStorageKeys` altogether eventually).